### ISSUE TITLE
[breaking change] try-ports is detrimental to good engineering

### ISSUE THREAD

**Issue by gcb** [dart-lang/sdk#54776]
### Change Intent

try-ports logic on `dart devtools` is detrimental and cause more issues on build systems than convenience (not sure who is the target. developers who run devtools one after another and forget them in the background? do we want to foster this? I honestly tried to understand when/why it was added but there was two migrations on git where this code was moved that made history hard to follow)

### Justification

This logic goes against every single other daemon behaviour. (i.e. user specify `--port=8000` but daemon listen on port `8001`, or `43993` or whatever other port. See bellow).

Also, probably a security issue somewhere. (e.g. deploy to stage where port 8000 is only accessible via office vpn and 8001 is connected to the your http cache egress, and devtools decide to start on port 8001).

If the caller have logic to handle that (i.e. read the command output and attach to a different port) then it should explicitly use that `--try-ports` param and be happy. (or `--port=0` see bellow)

Most daemons respect port parameters. It is counter-intuitive to start a webserver at a random port and search for it. If you use the `--port` command, and the port is already used, the common behaviour of most servers/daemons is to fail. I cannot think of other servers that does this by default. Specially undocumented like this.

Extra points:

- it's confusing that try-ports needs `1` to respect the value of `--port`. More confusing that the sensible value zero results in dart using a random port, ignoring `--port=8000` entirely. crazy.
- This eccentric and non-standard behaviour is not documented on most help text/online guide/`https://docs.flutter.dev/tools/devtools/cli`/etc. Only verbose help (`dart help -v <topic>`) for devtools will mention it exists! I will be honest, I didn't even know there was verbose help until I looked into a patch for this and saw the hidden param docs in the source. I expected help would have helped me.
- still on help text, the hidden help for this param is missing all the edge cases (e.g. `0` means use random port; or if `--port` failed because of permission like using port 80 as regular user, use random port; etc)
- The entire functionality of this (using unknown ports and figuring out by parsing output) is already handled by `--port=0`, which is already mentioned on the regular help for devtools just fine.

### Impact

More secure deployments. Saner server startup. Predictable behavior. Less silent failures on CI because server decided on a port on its own instead of failing.

Some program which already reads the output of `dart devtools` and parse it to get the port would fail *if also using `--port`*. If not using port it would not notice any change.

### Mitigation

set the default port to 0 (use any port).

Users who do not care about port and would be happy with random port will not even notice the change.

Better yet, `--port` could accept ranges (the implementation would be much simpler than what is there today for try-ports) and default to `--port=9100-9110`

**Comment by bkonyi** [comment]
It looks like this behavior was added in https://github.com/flutter/devtools/commit/87a9883765c08ad5f5a12076927a3ce424636e45 and the default value of `--try-ports` was set to 10 in https://github.com/flutter/devtools/pull/2458. 

I have to agree that this behavior is strange. I understand it's an annoyance to run `dart devtools` only to fail to bind to the default port, so this seems like it was a way to make things less annoying for developers on the CLI. I'm not sure this should be the default behavior, especially if `--port` is specified. If a user is specifying a specific port and DevTools can't be served on that port, the command should fail. cc @kenzieschmoll for their input.

I don't think this port selection behavior is much more of a security concern than providing `--port=0` and binding to a random port that may already be forwarded for another reason.

> try-ports logic on dart devtools, and its defaulting to 10, is detrimental and cause more issues on build systems than convenience.

Out of curiosity, how is this causing issues with build systems? I can't think of a scenario where the Dart VM service or DevTools should be enabled in that context.

**Comment by gcb** [comment]
> https://github.com/flutter/devtools/issues/1037

thanks for the history! Still don't see a valid use case though.

> https://github.com/flutter/devtools/issues/2305

Ironically asks for random ports, like you would get with ports=0.

> I have to agree that this behavior is strange. I understand it's an annoyance to run dart devtools only to fail to bind to the default port, so this seems like it was a way to make things less annoying for developers on the CLI. I'm not sure this should be the default behavior, especially if --port is specified. If a user is specifying a specific port and DevTools can't be served on that port, the command should fail. cc @kenzieschmoll for their input.

That's how I see it also. If `--port` is specified it should be honored. Everything else is better served by `--port=0` which can be the default and the `--try-ports` is never needed for any case. And I still fail to see who benefits from having tons of devtools lying around using the ports.

> I don't think this port selection behavior is much more of a security concern than providing --port=0 [...]

Correct, but then you are opt-in (if we don't make that the default. I personally think `--port=9100` being the default is better, and `--port=0` if you want any port. But `--port=0` might be a good middle ground so you guys can avoid a flood of tickets until a major version?).
Also, having a random port is better than sequential from a security POV, because sequential ports are more likely to be open by firewalls, being guessed, etc.
I am assuming the use case that don't care about port just use `--launch-browser` and never pay any attention to the port.

> Out of curiosity, how is this causing issues with build systems? 

enabling debug ports is very common to measure performance and code coverage in most languages. and opening a port range on the firewall is also common. so those are usually close together and with this 'feature' are a recipe for disaster.

A late functional stage build had a port used by something transient, devtools picked another random port. Much later on, things failed. The original port was not even used at that point, so we spent a lot of time hunting firewall and network issues and bad configs, when we were listening on the wrong port. Which then also caused some wasted time trying to figure out why the log said port X and configs said port Y and it worked everywhere but on that build step with the same config files. Again, remember I had to look at sources with the intent to suggest a patch before i even found out about `help -v` which is the only place with some documentation on this.

**Comment by bkonyi** [comment]
> That's how I see it also. If `--port` is specified it should be honored. Everything else is better served by `--port=0` which can be the default and the `--try-ports` is never needed for any case. And I still fail to see who benefits from having tons of devtools lying around using the ports.

There was probably a stronger argument back in the day when DevTools wasn't part of the Dart SDK and it had to be launched via `dart pub global run devtools`, so more users would have encountered issues with existing DevTool instances holding port 9100. It's not a strong argument, but it could definitely be a bit of a pain to find and kill the existing instances, especially if you're not as experienced.

> Also, having a random port is better than sequential from a security POV, because sequential ports are more likely to be open by firewalls, being guessed, etc.

Yeah, that's a good point.

> enabling debug ports is very common to measure performance and code coverage in most languages. and opening a port range on the firewall is also common. 

For sure. However, these workflows shouldn't be running `dart devtools` as tooling can query the VM service directly for this information programmatically without the need for a DevTools connection.

> A late functional stage build had a port used by something transient, devtools picked another random port. Much later on, things failed. The original port was not even used at that point, so we spent a lot of time hunting firewall and network issues and bad configs, when we were listening on the wrong port. Which then also caused some wasted time trying to figure out why the log said port X and configs said port Y and it worked everywhere but on that build step with the same config files. Again, remember I had to look at sources with the intent to suggest a patch before i even found out about `help -v` which is the only place with some documentation on this.

Would you mind sharing some high-level details about what you're building? 

For standalone Dart applications, we serve DevTools directly at the VM service URI, so this behavior can't happen. If Flutter is doing this when using the `flutter` tool, it probably shouldn't be as it reduces determinism of the tool unexpectedly.

**Comment by DanTup** [comment]
As I remember, part of the reason for using a fixed port was this:

https://github.com/flutter/devtools/issues/1037

> Now that we might show notifications, the user will get a notification permissions prompt for each port number we spawn on, so we should have an option to try 9100 and increase the number each time we fail to bind (up to some maximum).

Before we had DevTools embedded, it would also open in an external browser. If you kept clicking the Inspector icon on the toolbar (because the window was lost behind your IDE), we would show a browser notification "Hey, it's already open over here". Those notifications require permissions, and a changing port means the user would have to keep approving them (or more likely, the notification would not be seen).

With a default and `--try-ports`, it would hopefully keep things on a small number of ports for the purpose of those permissions (and perhaps bookmarks, etc.).

Given most use of DevTools is now embedded (and I'm not sure if we even still try to send notifications), I don't think this matters much from VS Code's perspective and it could just be random. (I'm less sure about implications of that for the CLI though).

**Comment by gcb** [comment]
> Would you mind sharing some high-level details about what you're building (e.g., a Dart or Flutter application)?
> 
> For standalone Dart applications, we serve DevTools directly at the VM service URI, so this behavior can't happen. If Flutter is doing this when using the flutter tool, it probably shouldn't be as it reduces determinism of the tool unexpectedly.

It is a flutter application but we are exploring calling dart devtools for some things.

Granted, we have a weird setup (big reason we are using flutter). We have volunteers/contributors in places where computers/storage/bandwidth are expensive and cannot run things like android studio, or even remote intellij (let alone an ios emulator). Se we provide a way to code against a vm where we host flutter/dart and provide web builds and then remote emulators. Those devs mostly have SSH and HTTPS, and nothing much else besides a text editor and a browser. I'm mostly working backwards from all the plugins for intellij/VS/webdev pkg.

**Comment by kenzieschmoll** [comment]
> If a user is specifying a specific port and DevTools can't be served on that port, the command should fail. cc @kenzieschmoll for their input.

Agreed. 

However, it is also annoying for users to run `dart devtools` and have it fail because it defaults to a specific port that is occupied. So whatever solution we end with, I think we should avoid making users pick a specific port in order to start DevTools (assuming the default port fails). We also have several tools that run `dart devtools`, and having those processes fail more aggressively would not be good. 

Perhaps that means we should default to `--port=0` and allow users to override with `--port=X`.

**Comment by gcb** [comment]
Btw a new option if you're afraid of backward compatibility should be fine too.
